### PR TITLE
bugfix: fix tpcds_logical_q8 ambiguous name.

### LINKED ISSUE #5334: tpcds_logical_q8 ambiguous name.
**Describe the bug**
`ambiguous name` problem due to q8

```
Error: Context("type_coercion", Internal("Optimizer rule 'type_coercion' failed due to unexpected error: Schema error: Ambiguous reference to unqualified field 'ca_zip'"))
```

In q8, there is a sql like following:
```sql
select project x as ca_zip
intersect
select ca_zip from a1
```
-> 
```sql
(project x as ca_zip) left semi join (a1) on "ca_zip = a1.ca_zip".
```

**To Reproduce**
After set `pub skip_failed_rules: bool, default = false`, we can find it.

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #5334.
Part of #4685.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

This PR fix tpcds_logical_q8.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/common/src/dfschema.rs [hunk 1]
```diff
@@ -284,12 +284,29 @@ impl DFSchema {
...
+                //             Join on id = b.id
+                // Project a.id as id   TableScan b id
+                // In this case, there isn't `ambiguous name` problem. When `matches` just contains
+                // one field without qualifier, we should return it.
+                let fields_without_qualifier = matches
+                    .iter()
+                    .filter(|f| f.qualifier.is_none())
+                    .collect::<Vec<_>>();
+                if fields_without_qualifier.len() == 1 {
+                    Ok(fields_without_qualifier[0])
```
Comment (alamb) [inline datafusion/common/src/dfschema.rs:300]: You might be able to write this using `find` instead

Something like

```suggestion
                if let Some(fields_without_qualifier) = matches
                    .iter()
                    .find(|f| f.qualifier.is_none()) {
                      Ok(fields_without_qualifier[0])
```

**Review by alamb — COMMENTED** [review]
This looks ok to me @jackwener  -- but I think we should have some sort of test that passes with this change and fails without it. The rationale being that without such coverage this logic may get lost during a refactoring or some other code change and we wouldn't know

File: datafusion/common/src/dfschema.rs (hunk 1 above)
Comment (jackwener) [reply datafusion/common/src/dfschema.rs:300]: When `fields_without_qualifier` size > 1, we should still return a Error.
So we shouldn't use `find`

**Comment by jackwener** [comment]
> but I think we should have some sort of test that passes with this change and fails without it. The rationale being that without such coverage this logic may get lost during a refactoring or some other code change and we wouldn't know because no test would fail

Thank you @alamb , I add a new integration-test for this issue.
Before this PR, it will fail. it will work after this PR.

File: datafusion/common/src/dfschema.rs (hunk 1 above)
Comment (alamb) [reply datafusion/common/src/dfschema.rs:300]: Makes sense :thumbsup:

**Review by alamb — APPROVED** [review]
Thanks @jackwener !

**Pull request merged**